woocommerce/woocommerce
# Support strict SQL modes in HPOS migration that dont allow zero date

### Submission Review Guidelines:

- I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
- I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/). 
- Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:


For many shops, MySQL/MariaDB server could be running the in [the `NO_ZERO_DATE` mode](https://dev.mysql.com/doc/refman/8.0/en/sql-mode.html#sqlmode_no_zero_date), in which case, we won't be able to insert data into HPOS tables when source metadata does not exist and destination type is datetime. The reason for this, is we pass every value before inserting via `$wpdb->prepare`, which [converts null values into empty string](https://core.trac.wordpress.org/ticket/12819).

To fix this, we will change our logic slightly, so that if the source is null, then we won't run the output through wpdb->prepare, instead will pass the hardcoded NULL value.

Closes #38320.



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

This bug does not present itself in UX, and also only on certain MySQL configurations. So we will use the shell, and force MySQL config that we need:

1. With posts table as authoritative and sync off, create a new order (using UI for this is fine). Let's say the ID of this order is XXX.
2. Open a wp shell, and test using the following command
```php

// First lets force the MySQL config to what we want.
global $wpdb;

$wpdb->query( "SET sql_mode = 'TRADITIONAL'" );

// Let's test that our config was set as expected, and 0 dates are not allowed.

use Automattic\WooCommerce\Internal\DataStores\Orders\OrdersTableDataStore;

$orders_table = OrdersTableDataStore::get_orders_table_name();

// this should return an error.
$result = $wpdb->query( "INSERT INTO $orders_table (date_created_gmt) VALUES ('0000-00-00 00:00:00')" ); 

assert( ! $result ); 

// Now we will try to migrate an order.
// Order ID from step 1.
$order_id = XXX; 
// forcing null value
delete_post_meta( $order_id, '_date_paid' ); 

use Automattic\WooCommerce\Database\Migrations\CustomOrderTable\PostsToOrdersMigrationController;

$sut = wc_get_container()->get( PostsToOrdersMigrationController::class );

 // should not raise an exception.
$sut->migrate_order( $order_id );

// should be empty
$errors = $sut->verify_migrated_orders( array( $order_id ) );

// should be true
assert( 0 === count($errors) );
```
